cinema/views.py

```diff
@@ -1,45 +1,110 @@
-from rest_framework.decorators import api_view
+from django.http import Http404
+from rest_framework.request import Request
 from rest_framework.response import Response
-from rest_framework import status
+from rest_framework import generics, status, mixins
 
-from django.shortcuts import get_object_or_404
+from rest_framework.views import APIView
+from rest_framework.viewsets import GenericViewSet, ModelViewSet
 
-from cinema.models import Movie
-from cinema.serializers import MovieSerializer
+from cinema.models import Movie, Genre, Actor, CinemaHall
+from cinema.serializers import (
+    ActorSerializer,
+    CinemaHallSerializer,
+    GenreSerializer,
+    MovieSerializer,
+)
 
 
-@api_view(["GET", "POST"])
-def movie_list(request):
-    if request.method == "GET":
-        movies = Movie.objects.all()
-        serializer = MovieSerializer(movies, many=True)
+class GenreList(APIView):
+    def get(self, request: Request) -> Response:
+        genres = Genre.objects.all()
+        serializer = GenreSerializer(genres, many=True)
         return Response(serializer.data, status=status.HTTP_200_OK)
 
-    if request.method == "POST":
-        serializer = MovieSerializer(data=request.data)
+    def post(self, request: Request) -> Response:
+        serializer = GenreSerializer(data=request.data)
+
         if serializer.is_valid():
             serializer.save()
-            return Response(serializer.data, status=status.HTTP_201_CREATED)
-
+            return Response(
+                serializer.data,
+                status=status.HTTP_201_CREATED
+            )
         return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
 
 
-@api_view(["GET", "PUT", "DELETE"])
-def movie_detail(request, pk):
-    movie = get_object_or_404(Movie, pk=pk)
+class GenreDetail(APIView):
+    def get_object(self, pk: int) -> Response:
+        try:
+            return Genre.objects.get(id=pk)
+        except Genre.DoesNotExist:
+            raise Http404
```

Comment:
I just kinda didn't know what to provide, since we need to keep it dry, but also we need to follow workflow, and it's kinda bad, when you should use low-lvl methods for some types and more high for others. You just totally miss idea on some point and add weird things )